Multithreaded vote processor

pwojcikdev opened the pull request:
Changes vote processor to use multiple threads. While processing of votes inside the election container is still done under exclusive lock, this should help amortize time spent checking vote signatures.

pwojcikdev committed:
Multithreaded vote processor

pwojcikdev committed:
Configurable batch size

gr0vity-dev commented:
I tested this on a local network, and somehow the multi-threaded approach behaves inferior to the current develop branch.
<img width="801" alt="Screenshot 2024-05-03 at 20 13 26" src="https://github.com/nanocurrency/nano-node/assets/85646666/c772833c-c51a-4472-923a-597d7bc7b36a">

Here are the [grafana stats](https://nl-nodestats.bnano.info/d/QjjDzXZSk/pwo-overview?orgId=1&var-Job=nanolab_e3b0c442&var-Instance=nl_pr1&var-Instance=nl_pr2&var-Instance=nl_pr3&var-Instance=nl_pr4&var-stats_type=vote_processor_overfill&var-stats_type=bootstrap_server_overfill&var-stats_type=blockprocessor_overfill&var-stats_detail=All&var-stats_dir=All&var-stats_objects_l1=All&var-Blocks=cemented&var-Blocks=count&var-message_type=All&var-Rate=30s&from=1714759240687&to=1714759656710)

pwojcikdev committed:
Fix notify

pwojcikdev commented:
That's interesting, thanks for testing @gr0vity-dev. This would imply that the overhead of lock contention is the largest factor here. I fixed a small bug with thread notifications, but I don't think it should have a significant impact. It would be great if you could run your tests with various settings for `batch_size` and `threads`. For now I modified the default config to use a single thread by default.

clemahieu commented:
A frame graph would be good at identifying what's causing the slowdown, seems odd that one thread would be the fastest for an explicitly CPU-bound task.

pwojcikdev committed:
Improve `fair_queue::next_batch ()`

pwojcikdev committed:
Container info

pwojcikdev commented:
I received the flamegraph and was able to identify the bottleneck. It's now fixed and multithreaded version seems to perform better than single threaded, as expected.

pwojcikdev committed:
Further optimize `fair_queue::size ()` calculation

pwojcikdev committed:
Cleanup

gr0vity-dev commented:
I did another side by side run between develop and the improved multi-threaded vote processor
<img width="884" alt="Screenshot 2024-05-05 at 19 59 42" src="https://github.com/nanocurrency/nano-node/assets/85646666/a2b8979d-6e12-4356-978c-030c8d74f7ca">

[Here](https://nl-nodestats.bnano.info/d/nano-overview-histograms-extended/nano-stats-and-counters-histograms-extended?orgId=1&from=1714931304117&to=1714931745943&var-Job=nanolab_e3b0c442&var-Instance=All&var-stats_type=All&var-stats_detail=All&var-stats_dir=All&var-stats_objects_l1=All&var-Blocks=All&var-message_type=All&var-Rate=30s&var-Step=1s&var-samples_detail=All&var-rpc_method=All&var-priority_bucket=All) are the detailed grafana stats
